Automattic/wp-calypso
# JP Onboarding: Make saveJetpackOnboardingSettings() generic, use to perform JPC-auth'd requests

Make `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` more generic in order to make them work without JPO auth tokens when connected. We now simply pass the JPO authentication token as an argument to these actions; we omit them if we're already connected to WP.com (per JPC). This means that this PR isn't simply a refactor: Instead, it allows to run the JPO flow without JPO token when connected to WP.com.

Some architectural choices:
* I was trying to re-implement `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` by just hard-wiring the token to more generic `saveJetpackSettings()` and `requestJetpackSettings() actions, respectively. However, that requires obtaining the token via `getState()`, i.e. it would cause `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` to be thunks. However, we cannot use the `getRequest()` selector (which we use to keep track of `isLoading` state) with thunks -- it only works with normal Redux action objects. Thus, I opted to just provide  `saveJetpackSettings()` and `requestJetpackSettings()` and leave it to the consumer to provide the onboarding token.
* This actually works quite nicely in the `<Main />` component's `connect()` clause, where we can centralize the logic that determines whether the site is connected and either omit or provide the onboarding token accordingly.

Note that this makes reducers and actions in `state/jetpackOnboarding` somewhat redundant with what is found in `state/jetpack/settings`. However, I opted to keep "ours" (`state/jetpackOnboarding`)  since it uses the data layer (and is arguably more elegant 😉). In a subsequent PR, I'd like to change consumers of `state/jetpack/settings` to use "ours" instead.

Also note that the directory is still called `state/jetpackOnboarding`, even though it can now generically work with JP settings.

Finally, note that `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` will be renamed to `saveJetpackSettings()` and `requestJetpackSettings()`, respectively, in the follow-up PR that will deduplicate `state/jetpack/settings` (alongside action types etc). As per discussion with @tyxla, I'm not doing that as part of this PR in order to somewhat separate PR concerns, and to keep diff size down.

To test:
This requires some thorough testing, since it touches almost all of JPO state.
* Verify that JPO (while _not connected_ to WP.com) still works.
* In particular, verify that WooCommerce install retries still work on a slow host (try Pressable) (this is to verify the `retryOrAnnounceSaveFailure()` function still works).
* Also, pay attention to the summary page, and whether checkmarks/crosses/spinners are shown correctly for each individual step.
* Verify that JPO also works while connected to WP.com.

Fixes #22578 🙂
